README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ggepi

[![Travis-CI Build
Status](https://travis-ci.org/lwjohnst86/ggepi.svg?branch=master)](https://travis-ci.org/lwjohnst86/ggepi)
[![lifecycle](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)

The main goal of the ggepi package is to provide ggplot2 layers for
common statistical graphics (at least in chronic disease epidemiology
and at least what I use often). The ggepi package is as hands off as
possible. This package assumes what you want to plot is what you want to
plot, and lets you have full control over how the plot of the results
looks like.

# Installation

Install the developmental version using:

``` r
install.packages("devtools")
devtools::install_github("lwjohnst86/ggepi")
```

# Usage

Check out the website documents for more details.


NEWS.md
# ggepi 0.0.1.9000

* Added `geom_corr_circle` layer (#3) for PLS and PCA type results.

# ggepi 0.0.1

* Created a [pkgdown](https://hadley.github.io/pkgdown/index.html) site for the package.
* Added a introduction vignette describing the package.
* Created a standard, single function to create a typical plot (`view_estci`).
* Created a custom `geom_estci` ggplot2 layer that visualizes regression results.
* Added a `NEWS.md` file to track changes to the package.


R/utils.R

#' Geometry for circles in ggforce.
#'
#' See [ggforce::geom_circle()].
#'
#' @name GeomCircle
#' @rdname GeomCircle
#' @keywords internal
#' @export
#' @importFrom ggforce GeomCircle
NULL

# From ggplot2's utilities.
"%||%" <- ggplot2:::"%||%"

#' Convert continuous p-values into a factor based on the given split.
#'
#' @param p.value Vector of p-values.
#' @param split_by Where to split the p-values up by. Default is the <0.001,
#'   <0.01, <0.05, and all others are >0.05.
#'
#' @return A factor vector of p-value splits.
#' @export
#'
#' @examples
#'
#' pvals <- c(0.001, 0.0123, 0.123, 0.009, 0.2145, 0.987, 0.05, 0.049, 0.051)
#' discrete_pvalue(pvals)
#' discrete_pvalue(pvals, split_by = c(0.05))
#' discrete_pvalue(pvals, split_by = c(0.01))
#' discrete_pvalue(pvals, split_by = c(0.01, 0.05))
#' discrete_pvalue(pvals, split_by = c(0.05, 0.001, 0.01))
discrete_pvalue <- function(p.value, split_by = c(0.001, 0.01, 0.05)) {
    stopifnot(is.numeric(p.value), is.numeric(split_by))
    split_by <- sort(split_by)
    sig_pval_labels <- paste0("<", split_by)
    nonsig_pval_labels <- paste0(">", split_by[length(split_by)])
    pval_labels <- c(sig_pval_labels, nonsig_pval_labels)
    pval_discrete <- cut(
        p.value,
        breaks = c(-Inf, split_by, Inf),
        labels = pval_labels,
        ordered_result = TRUE
    )

    factor(pval_discrete, levels = rev(levels(pval_discrete)))
}


R/geom_estci.R
#' Horizontal point and line range with vertical center line.
#'
#' Useful to visualize results from regression type analyses, as it shows the
#' estimate, confidence interval, and optionally use the value of the p.value
#' to highlight significant associations. A vertical line is included
#'
#' @inheritParams ggplot2::layer
#' @inheritParams ggplot2::geom_pointrange
#' @param height Add ends to the confidence intervals.
#' @param center.linetype The linetype for the center line.
#' @param center.linecolour Line colour for the center line.
#' @param center.linesize Line size for the center line.
#' @param ci.linesize Line size for the confidence interval lines.
#'
#' @return Adds a ggplot2 geom layer.
#' @export
#'
#' @examples
#'
#' library(ggplot2)
#' library(broom)
#'
#' fit <- lm(Fertility ~ 0 + Catholic + Agriculture + Examination +
#' Education + Infant.Mortality, data = swiss)
#' fit <- tidy(fit, conf.int = TRUE)
#' fit <- transform(fit, model = "non-log", p.value = discrete_pvalue(fit$p.value))
#'
#' p <- ggplot(fit, aes(x = estimate, y = term, xmin = conf.low, xmax = conf.high))
#' p + geom_estci()
#'
#' p + geom_estci(aes(xintercept = 1.1), center.linecolour = "red")
#' p + geom_estci(aes(size = p.value), linetype = "dotted")
#' p + geom_estci(aes(colour = p.value, size = p.value), linetype = "dotted")
#' p + geom_estci(aes(colour = p.value, size = p.value), linetype = "dotted") +
#'   scale_colour_grey(start = 0.75, end = 0)
#' p + geom_estci(aes(size = p.value, alpha = p.value), linetype = "dotted")
#' p + geom_estci(aes(size = p.value, alpha = p.value, colour = p.value))
#' p + geom_estci(aes(alpha = p.value), linetype = "dashed",
#' center.linetype = "solid")
#'
#' p + geom_estci(aes(alpha = p.value, xintercept = 1),
#' colour = "blue", linetype = "dashed", center.linetype = "solid")
#'
#' p + geom_estci(aes(alpha = p.value, xintercept = 1), center.linesize = 1.5)
#' p + geom_estci(center.linesize = 0.25, height = 1, fatten = 2)
#' p + geom_estci(center.linesize = 2, height = 0.5, fatten = 8)
#' p + geom_estci(ci.linesize = 3)
#' p + geom_estci(aes(size = p.value, colour = p.value), fatten = 2)
#'
#' fit_log <- lm(log(Fertility) ~ 0 + Catholic + Agriculture + Examination +
#'     Education + Infant.Mortality, data = swiss)
#' fit_log <- tidy(fit_log, conf.int = TRUE)
#' fit_log <- transform(fit_log, model = "log",
#'     p.value = discrete_pvalue(fit_log$p.value))
#' two_fits <- rbind(fit, fit_log)
#'
#' p <- ggplot(two_fits, aes(x = estimate, y = term, xmin = conf.low, xmax = conf.high))
#' # It might be possible to show groups with 'dodging', but it is currently in development.
#' # p + geom_estci(aes(group = model, colour = model), position = position_dodge(width = 0.3))
#' p + geom_estci()
#'
#' p <- ggplot(two_fits, aes(x = estimate, y = term, xmin = conf.low, xmax = conf.high))
#' p + geom_estci() + facet_grid(~ model)
#'
geom_estci <- function(mapping = NULL,
                       data = NULL,
                       stat = "identity",
                       position = "identity",
                       ...,
                       height = NA,
                       fatten = 4,
                       na.rm = FALSE,
                       show.legend = NA,
                       center.linetype = "dashed",
                       center.linecolour = "black",
                       center.linesize = 0.5,
                       ci.linesize = 0.5,
                       inherit.aes = TRUE) {

    layer(
        data = data,
        mapping = mapping,
        stat = stat,
        geom = GeomEstci,
        position = position,
        show.legend = show.legend,
        inherit.aes = inherit.aes,
        params = list(
            fatten = fatten,
            height = height,
            na.rm = na.rm,
            center.linetype = center.linetype,
            center.linecolour = center.linecolour,
            center.linesize = center.linesize,
            ci.linesize = ci.linesize,
            ...
        )
    )
}

#' @rdname ggepi-geoms
#' @format NULL
#' @usage NULL
#' @export
GeomEstci <- ggproto(
    "GeomEstci",
    Geom,

    default_aes = aes(
        colour = "black",
        size = 0.5,
        linetype = 1,
        height = NA,
        alpha = NA,
        shape = 19,
        fill = NA,
        stroke = 0.5,
        xintercept = 0
    ),
    required_aes = c("y", "x", "xmin", "xmax", "xintercept"),

    draw_key = draw_key_point,

    setup_data = function(data, params) {
        data$height <- data$height %||%
            params$height %||% (resolution(data$y, FALSE) * 0.9)

        data$xintercept <- data$xintercept %||%
            params$xintercept %||% 0

        data$center.linetype <- data$center.linetype %||%
            params$center.linetype %||% "dotted"

        data$center.linecolour <- data$center.linecolour %||%
            params$center.linecolour %||% "black"

        data$center.linesize <- data$center.linesize %||%
            params$center.linesize %||% 0.5

        data$ci.linesize <- data$ci.linesize %||%
            params$ci.linesize %||% 0.5

        transform(
            data,
            ymin = y - height / 2,
            ymax = y + height / 2,
            height = NULL
        )
    },

    draw_group = function(data,
                          panel_params,
                          coord,
                          height = NULL,
                          fatten = NA,
                          center.linetype = NA,
                          center.linecolour = NA,
                          center.linesize = NA,
                          ci.linesize = NA) {

        data_centerline <- transform(
            data,
            linetype = data$center.linetype,
            alpha = NA,
            colour = data$center.linecolour,
            size = data$center.linesize
        )

        data_ci <- transform(
            data,
            size = data$ci.linesize
        )

        ggplot2:::ggname("geom_estci", grid::grobTree(
            GeomVline$draw_panel(data_centerline, panel_params, coord),
            GeomErrorbarh$draw_panel(data_ci, panel_params, coord, height = height),
            GeomPoint$draw_panel(transform(data, size = size * fatten), panel_params, coord)
        ))
    }
)


R/view_estci.R

#' Standard plot for estimates and confidence interval.
#'
#' Typical figure for examining estimates and confidence intervals from
#' regression type analyses. Has standard features for the plot, but if you want
#' more customization, see the \code{\link{geom_estci}} function.
#'
#' @seealso geom_estci
#'
#' @param data Tidied dataframe from a linear regression, as from the function
#'   \link[broom]{tidy} from the broom package.
#' @param xterms Column that contains the predictor variables (ie. "term").
#' @param split_by Column to split the results up by (e.g. different models or
#'   different y terms).
#' @param ylab Y-axis label.
#' @param xlab X-axis label.
#' @param legend.title Legend title.
#' @param center.line Location for the center line.
#'
#' @return Plot of estimate and confidence interval.
#' @export
#'
#' @examples
#'
#' library(broom)
#' fit <- lm(Fertility ~ 0 + Catholic + Agriculture + Examination +
#'     Education + Infant.Mortality, data = swiss)
#' fit <- tidy(fit, conf.int = TRUE)
#' fit <- transform(fit, model = "non-log")
#'
#' view_estci(fit, xterms = "term")
#' view_estci(fit, xterms = "term", center.line = 1)
#'
#' fit_log <- lm(log(Fertility) ~ 0 + Catholic + Agriculture + Examination +
#'     Education + Infant.Mortality, data = swiss)
#' fit_log <- tidy(fit_log, conf.int = TRUE)
#' fit_log <- transform(fit_log, model = "log")
#' two_fits <- rbind(fit, fit_log)
#'
#' view_estci(two_fits, xterms = "term", split_by = "model", center.line = 1)
view_estci <-
    function(data,
             xterms = "term",
             split_by = NA,
             ylab = "Predictor terms",
             xlab = "Estimate and CI",
             legend.title = "P-value",
             center.line = 0) {

        needed_columns <-
            setdiff(stats::na.omit(
                c(
                    "estimate",
                    "conf.low",
                    "conf.high",
                    "p.value",
                    xterms,
                    split_by
                )
            ),
            names(data))

        if (length(needed_columns) != 0) {
            stop(
                "The columns (",
                paste(needed_columns, collapse = ", "),
                ") are missing from the dataset.",
                call. = FALSE
            )
        }

        data$p.value <- discrete_pvalue(data$p.value)

        p <-
            ggplot(
                data,
                aes_string(
                    x = "estimate",
                    y = xterms,
                    xmin = "conf.low",
                    xmax = "conf.high",
                    size = "p.value",
                    colour = "p.value"
                )
            ) +
            geom_estci(aes(xintercept = center.line), fatten = 2) +
            scale_color_grey(start = 0.75,
                             end = 0,
                             name = legend.title) +
            scale_size_discrete(name = legend.title, range = c(0.5, 2)) +
            theme_classic() +
            theme(strip.background = element_blank()) +
            labs(y = ylab, x = xlab)

        if (!is.na(split_by)) {
            p <- p + facet_grid(paste0("~", split_by), scales = "free")
        }

        return(p)
    }


R/geom_corr_circle.R
#' Correlation circle plot for dimensionality reduction techniques.
#'
#' This creates a ggplot2 layer that plots the correlation values of the
#' scores for components against the real values, as obtained from
#' dimensionality reduction methods. These methods include principal components
#' analysis and partial least squares.
#'
#' @inheritParams ggplot2::layer
#' @inheritParams ggplot2::geom_point
#' @param center.linetype,center.linecolour,center.linesize The line type,
#'   colour, and size for the two center lines.
#' @param inner.linetype,inner.linecolour,inner.linesize The line type,
#'   colour, and size for the inner circle line.
#' @param outer.linetype,outer.linecolour,outer.linesize The line type,
#'   colour, and size for the outer circle line.
#'
#' @return Adds a ggplot2 geom layer.
#' @export
#'
#' @examples
#'
#' library(ggplot2)
#' library(broom)
#' library(pls)
#' data(yarn)
#'
#' # Set up data to plot.
#' fit <- plsr(density ~ NIR, 6, data = yarn, validation = "CV")
#' fit <- cor(model.matrix(fit), scores(fit)[, 1:2, drop = FALSE])
#' fit <- as.data.frame(fit)
#' fit$Variables <- rownames(fit)
#' rownames(fit) <- NULL
#' colnames(fit)[1:2] <- c("Comp1", "Comp2")
#' fit$Groups <- as.factor(sample.int(2, nrow(fit), replace = TRUE))
#'
#' p <- ggplot(fit, aes(x = Comp1, y = Comp2))
#' p + geom_corr_circle()
#' p + geom_corr_circle(outer.linetype = "dotted")
#' p + geom_corr_circle(inner.linecolour = "blue")
#' p + geom_corr_circle(center.linesize = 1)
#' p + geom_corr_circle(center.linecolour = "grey50", size = 3)
#'
#' # If you want to remove the circle or center lines, use 0.
#' p + geom_corr_circle(center.linetype = 0)
#' p + geom_corr_circle(outer.linetype = 0)
#'
#' # With grouping
#' p + geom_corr_circle(aes(colour = Groups))
#' p + geom_corr_circle(aes(colour = Groups), size = 3) +
#' scale_colour_brewer()
#' p + geom_corr_circle(aes(alpha = Groups))
#' p + geom_corr_circle(aes(size = Groups))
#' p + geom_corr_circle(aes(shape = Groups), size = 2)
#'
#' # With facets
#' p + geom_corr_circle() +
#' facet_grid(~ Groups)
#'
geom_corr_circle <- function(mapping = NULL,
                       data = NULL,
                       stat = "identity",
                       position = "identity",
                       ...,
                       na.rm = FALSE,
                       show.legend = NA,
                       outer.linetype = "solid",
                       outer.linecolour = "black",
                       outer.linesize = 0.5,
                       inner.linetype = "dotted",
                       inner.linecolour = "black",
                       inner.linesize = 0.5,
                       center.linetype = "solid",
                       center.linecolour = "grey50",
                       center.linesize = 0.3,
                       inherit.aes = TRUE) {

    layer(
        data = data,
        mapping = mapping,
        stat = stat,
        geom = GeomCorrcircle,
        position = position,
        show.legend = show.legend,
        inherit.aes = inherit.aes,
        params = list(
            na.rm = na.rm,
            outer.linetype = outer.linetype,
            outer.linecolour = outer.linecolour,
            outer.linesize = outer.linesize,
            inner.linetype = inner.linetype,
            inner.linecolour = inner.linecolour,
            inner.linesize = inner.linesize,
            center.linetype = center.linetype,
            center.linecolour = center.linecolour,
            center.linesize = center.linesize,
            ...
        )
    )
}

#' @rdname ggepi-geoms
#' @format NULL
#' @usage NULL
#' @export
GeomCorrcircle <- ggproto(
    "GeomCorrcircle",
    Geom,

    default_aes = aes(
        colour = "black",
        size = 0.5,
        linetype = 1,
        alpha = NA,
        fill = NA,
        shape = 19,
        stroke = 0.5
    ),
    required_aes = c("y", "x"),

    draw_key = draw_key_point,

    setup_data = function(data, params) {

        data$outer.linetype <- data$outer.linetype %||%
            params$outer.linetype %||% "solid"
        data$outer.linecolour <- data$outer.linecolour %||%
            params$outer.linecolour %||% "black"
        data$outer.linesize <- data$outer.linesize %||%
            params$outer.linesize %||% 0.5

        data$inner.linetype <- data$inner.linetype %||%
            params$inner.linetype %||% "dotted"
        data$inner.linecolour <- data$inner.linecolour %||%
            params$inner.linecolour %||% "black"
        data$inner.linesize <- data$inner.linesize %||%
            params$inner.linesize %||% 0.5

        data$center.linetype <- data$center.linetype %||%
            params$center.linetype %||% "solid"
        data$center.linecolour <- data$center.linecolour %||%
            params$center.linecolour %||% "grey50"
        data$center.linesize <- data$center.linesize %||%
            params$center.linesize %||% 0.3

        data
    },

    draw_panel = function(data,
                          panel_params,
                          coord,
                          outer.linetype = NA,
                          outer.linecolour = NA,
                          outer.linesize = NA,
                          inner.linetype = NA,
                          inner.linecolour = NA,
                          inner.linesize = NA,
                          center.linetype = NA,
                          center.linecolour = NA,
                          center.linesize = NA
                          ) {

        outer_circle <- util_circle_data(1, npoints = nrow(data))
        data_outer_circle <- transform(
            data,
            group = -1,
            x = outer_circle$x,
            y = outer_circle$y,
            linetype = data$outer.linetype,
            alpha = NA,
            colour = data$outer.linecolour,
            size = data$outer.linesize
        )

        inner_circle <- util_circle_data(sqrt(1 / 2), npoints = nrow(data))
        data_inner_circle <- transform(
            data,
            group = -1,
            x = inner_circle$x,
            y = inner_circle$y,
            linetype = data$inner.linetype,
            alpha = NA,
            colour = data$inner.linecolour,
            size = data$inner.linesize
        )

        data_hline <- transform(
            data,
            group = -1,
            x = -1,
            y = 0,
            xend = 1,
            yend = 0,
            alpha = NA,
            linetype = data$center.linetype,
            colour = data$center.linecolour,
            size = data$center.linesize
        )
        data_hline <- unique(data_hline)

        data_vline <- transform(
            data,
            group = -1,
            x = 0,
            y = -1,
            xend = 0,
            yend = 1,
            alpha = NA,
            linetype = data$center.linetype,
            colour = data$center.linecolour,
            size = data$center.linesize
        )
        data_vline <- unique(data_vline)

        ggplot2:::ggname(
            "geom_corr_circle",
            grid::grobTree(
                GeomPoint$draw_panel(data, panel_params, coord),
                GeomSegment$draw_panel(data_hline, panel_params, coord),
                GeomSegment$draw_panel(data_vline, panel_params, coord),
                GeomPath$draw_panel(data_inner_circle, panel_params, coord),
                GeomPath$draw_panel(data_outer_circle, panel_params, coord)
            )
        )
    }
)

#         ggrepel::geom_text_repel(
#             data = fit,
#             aes(label = xvariables),
#             size = 2.5,
#             box.padding = 0.4,
#             segment.alpha = 0.3
#         )


util_circle_data <-
    function(radius = 1,
             center = c(0, 0),
             npoints = 100) {
        tt <- seq(0, 2 * pi, length.out = npoints)
        xvalues <- center[1] + radius * cos(tt)
        yvalues <- center[2] + radius * sin(tt)
        return(data.frame(x = xvalues, y = yvalues))
    }
